dpnp/dparray.pyx

```diff
@@ -45,19 +45,98 @@ import numpy
+        cdef dict iface
+        cdef _ArrayInterfaceData res_iface
+        cdef sycl_ctx
+        cdef char * iface_pointer
+        
+        if not hasattr(memory, "__sycl_usm_array_interface__"):
+            raise TypeError("Memory object does not expose __sycl_usm_array_interface__")
+        iface = memory.__sycl_usm_array_interface__
+        iface_ver = iface['version']
+        if (iface_ver != 1):
+            raise TypeError("Memory object exposes __sycl_usm_array_interface__ of version {}, version 1 is expected".format(iface_ver))
+        data_obj = iface.get('data', None)
+        if (not isinstance(data_obj, (tuple, list)) or len(data_obj) != 2):
+            raise TypeError("'data' entries in __sycl_usm_array_interface__ dictionary is not a tuple of length 2")
+        
+        iface_pointer = <char *>(<Py_ssize_t>data_obj[0]);
+        writeable = (<int>data_obj[1])
+        typestr = iface.get('typestr')
+        dtype = numpy.dtype(typestr)
+        if (dtype.kind == 'V'): # follow NumPy here
+            typedescr = iface.get('descr', None)
+            dtype = numpy.dtype((typestr, typedescr))
+        shape = iface['shape']
+        if not isinstance(shape, (list, tuple)):
+            raise TypeError("'shape' entry in __sycl_usm_array_interface__ dictionary is not a list or a tuple")
+        nd = len(shape)
+        strides = iface.get('strides', None)
+        if (strides is not None):
+            if not isinstance(strides, (list, tuple)) or len(strides) != nd or nd == 0:
+                raise TypeError("'shape' entry in __sycl_usm_array_interface__ dictionary is not a list or a tuple")
+            
+        offsets = iface.get('offsets', 0)
+        syclobj = iface.get('syclobj')
+        if not isinstance(syclobj, (dpctl.SyclQueue, dpctl.SyclContext)):
+            raise TypeError("'syclobj' entry in __sycl_usm_array_interface__ dictionary is not "
+                            "dpctl.SyclQueue or dpctl.SyclContext")
+        if isinstance(syclobj, dpctl.SyclQueue):
+            iface_queue = <c_dpctl.SyclQueue> syclobj
+            usm_type = c_dpctl_mem._Memory.get_pointer_type(
+                <c_dpctl.DPPLSyclUSMRef>iface_pointer, iface_queue.get_sycl_context())
+            if (usm_type != b'shared'):
+                raise TypeError("Expecting USM shared memory.")
+        else:
+            # Obtain device from pointer and context
+            sycl_ctx = <c_dpctl.SyclContext> syclobj
+            sycl_dev = c_dpctl_mem._Memory.get_pointer_device(<c_dpctl.DPPLSyclUSMRef>iface_pointer, ctx)
+            usm_type = c_dpctl_mem._Memory.get_pointer_type(<c_dpctl.DPPLSyclUSMRef>iface_pointer, ctx)
+            if (usm_type != b'shared'):
+                raise TypeError("Expecting USM shared memory.")
+            # Use context and device to create a queue to
+            # be able to copy memory
+            iface_queue = c_dpcl.SyclQueue._create_from_context_and_device(sycl_ctx, sycl_dev)
```

Comment:
it is not quite clear why you need to `create` a queue. Use the `queue` to copy data.

Reply:
This is the case when `__sycl_usm_array_interface__` is used to shared data allocated without a queue, using `sycl::malloc(size_t n, sycl::context, sycl::device, sycl::usm::kind)`, and only records the `context`.

We then use `sycl::get_pointer_device(ptr, ctx)` to look up the device on which memory was allocated, and need to create a queue against which operations on this memory can be performed.